### PR TITLE
lints: lint for upcoming Apple max cert lifetime policy.

### PR DESCRIPTION
**Pull request by cpu**
A new `e_tls_server_cert_valid_time_longer_than_398_days` lint is added for the Apple source category (presently named `lint.AppleCTPolicy`, see https://github.com/zmap/zlint/issues/418).

This lint returns an error lint result if a server-auth certificate issued after Sept 1st, 2020 has a lifetime > 398 days. The lifetime is calculated as Apple specifies, e.g. "398 days is measured with a day being equal to 86,400 seconds.".

A warning result is returned if a certificate issued after Sept 1st, 2020 has a lifetime > 397 days and < 398 days. This matches Apple's SHOULD-equivalent recommendation to use a validity period <= 397 days in length.

See https://support.apple.com/en-us/HT211025 for more information.

Resolves https://github.com/zmap/zlint/issues/407

### CODE REVIEW

File: v2/lints/apple/lint_e_server_cert_valid_time_longer_than_398_days.go [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+	errEndDate := c.NotBefore.Add(maxValidity)
+	warnEndDate := c.NotBefore.Add(warnValidity)
+
+	if c.NotAfter.After(errEndDate) {
+		return &lint.LintResult{Status: lint.Error}
+	} else if c.NotAfter.After(warnEndDate) {
+		return &lint.LintResult{
+			Status: lint.Notice,
+			Details: "Apple recommends that certificates be issued with a maximum " +
+				"validity of 397 days.",
```
[the remark this answers is not in the document]
Comment (sleevi) [reply v2/lints/apple/lint_e_server_cert_valid_time_longer_than_398_days.go]: You called the variable warnEndDate, but set the warning level to Notice.

This does seem to raise to the level of SHOULD (RFC 2119 has SHOULD and RECOMMENDED as equal), and thus be a lint.Warn situation. It's very much in line with Apple's goal here - you SHOULD be 397 or less, unless you have a dang good reason and know it's danger zone if you can't math.

File: v2/testdata/eeServerCertValidEqual397.pem [hunk 2]
```diff
@@ -0,0 +1,10 @@
+-----BEGIN CERTIFICATE-----
```
Comment (sleevi) [inline v2/testdata/eeServerCertValidEqual397.pem]: Should these test files include the openssl dump (aka -text)

The other test certs seem to.

File: v2/lints/apple/lint_e_server_cert_valid_time_longer_than_398_days.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cpu) [reply v2/lints/apple/lint_e_server_cert_valid_time_longer_than_398_days.go]: > Why did you wrap this, but not line 64?

Missed it :-) Fixed across the file in 5d7b0d4.

> This does seem to raise to the level of SHOULD (RFC 2119 has SHOULD and RECOMMENDED as equal), and thus be a lint.Warn situation. 

Your argument here makes sense to me :+1: I changed to a `lint.Warn` and added a brief comment with rationale in cc5c7f0.

Thanks!

File: v2/testdata/eeServerCertValidEqual397.pem (hunk 2 above)
Comment (cpu) [reply v2/testdata/eeServerCertValidEqual397.pem]: I think I'm guilty of forgetting to do this with some frequency :sweat:

Fixed for this branch in 961cc9c. I cleaned up the utility script and will push a separate PR that adds the same text output to all the test files that are missing it in master.

**Comment by cpu** [comment]
> Looks good to me, but with a few nits

Thanks for the review @sleevi! I think it's ready for a second :mag: when you have a chance.

**Comment by zakird** [comment]
No need to block on me

**Comment by cpu** [comment]
Thanks all. I'll update the PR description and squash merge it momentarily.

**Pull request merged**